cli/lib/output-mapper.ts

```diff
@@ -104,23 +120,51 @@ export async function mapDimensions(this: Pick<Context, 'variables'>) {
     return value
   }
 
-  const { default: map } = await importDynamic('barnard59-base/map.js')
+  // We store non-observation quads in memory and inject them into the stream
+  // after all the observation-specific steps
+  const nonObservationQuads = $rdf.dataset()
+  this.variables.set('nonObservationQuads', nonObservationQuads)
 
-  return map(async (quad: Quad) => {
+  const handleQuad = async (quad: Quad) => {
     const { subject, predicate, object, graph } = quad
     const mappingTerm = getDimensionMapping(predicate)
     if (mappingTerm?.value) {
       if (object.equals(undef)) {
-        return $rdf.quad(subject, predicate, cube.Undefined, graph)
+        return [$rdf.quad(subject, predicate, cube.Undefined, graph)]
       }
 
       const mappedValue = await getMappedValue(mappingTerm.value, object)
       if (mappedValue?.termType === 'NamedNode') {
-        return $rdf.quad(subject, predicate, mappedValue, graph)
+        const predicateOriginal = $rdf.namedNode(predicate.value + '#_original')
+
+        nonObservationQuads.add($rdf.quad(predicateOriginal, rdf.type, cc.OriginalValuePredicate, graph))
+        nonObservationQuads.add($rdf.quad(predicateOriginal, schema.about, predicate, graph))
+
+        return [
+          $rdf.quad(subject, predicate, mappedValue, graph),
+          $rdf.quad(subject, predicateOriginal, object, graph),
+        ]
       }
     }
 
-    return quad
+    return [quad]
+  }
+
+  return flatMap(handleQuad)
```

Comment:
I'm trying to understand why the `flatMap` is required. You don't really need the `predicateOriginal` quad push right away. If you also add it to the in-memory dataset the map step can be much like before.

Reply:
By the way, doesn't the current implementation cause `#_original` to become a cube dimension in the cube. I don't see any changes in the `toCubeShape` etc

Reply:
It didn't seem to cause any issue. But I'll apply your suggestion.

Reply:
Implementation details aside, I think we need a test to verify that there are indeed no extra dimensions created in the cube constraint shape.

Reply:
By the way, storing all the original values in the `variable` will increase the memory footprint of the pipeline. Hopefully not too much...

Reply:
My stream-fu is not strong enough but I wonder if we could also have it as a stream, but separate from the rest of the steps, and merged in the final output